repose-aggregator/components/filters/header-normalization-filter/src/main/scala/org/openrepose/filters/headernormalization/HeaderNormalizationFilter.scala

```diff
@@ -71,22 +73,42 @@ class HeaderNormalizationFilter @Inject()(configurationService: ConfigurationSer
 
     val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
 
-    config find { target =>
+    configRequest find { target =>
       // find the first "target" config element that matches this request (if any)
       target.url.matcher(wrappedRequest.getRequestURI).matches &&
         (target.methods.contains(wrappedRequest.getMethod) || target.methods.contains(AllHttpMethods))
     } foreach { target =>
       // figure out which headers to remove, and remove them
       (target.access match {
         case WhiteList => wrappedRequest.getHeaderNamesScala.diff(target.headers)
-          // written like this to maintain the case-insensitive string comparisons - do not swap
-        case BlackList => target.headers.intersect(wrappedRequest.getHeaderNamesScala)
+        case BlackList => target.headers
       }).foreach(wrappedRequest.removeHeader)
 
-      metricsMeter.mark(s"${target.url}_${wrappedRequest.getMethod}")
+      metricsMeter.mark(s"${target.url}_${wrappedRequest.getMethod}_request")
     }
 
-    filterChain.doFilter(wrappedRequest, servletResponse)
+    val wrappedResponse = if (configResponse.isEmpty) None else Option(
+      new HttpServletResponseWrapper(servletResponse.asInstanceOf[HttpServletResponse], MUTABLE, PASSTHROUGH))
+
+    filterChain.doFilter(wrappedRequest, wrappedResponse.getOrElse(servletResponse.asInstanceOf[HttpServletResponse]))
+
+    configResponse find { target =>
+      // find the first "target" config element that matches this request (if any)
+      target.url.matcher(wrappedRequest.getRequestURI).matches &&
+        (target.methods.contains(wrappedRequest.getMethod) || target.methods.contains(AllHttpMethods))
+    } foreach { target =>
+      // figure out which headers to remove, and remove them
+      (target.access match {
+        // The Response Wrapper doesn't return a case-insensitive collection of Header Names at this time.
+        case WhiteList => (new TreeSet[String]()(caseInsensitiveOrdering) ++
+          wrappedResponse.get.getHeaderNames.asScala).diff(target.headers)
```

Comment:
For the Ordering approach, these lines become:
```
val headerNames = wrappedResponse.get.getHeaderNames.asScala.toSet
headerNames -- headerNames.intersect(target.headers)
```

or

```
val headerNames = wrappedResponse.get.getHeaderNames.asScala.toSet
headerNames.diff(headerNames.intersect(target.headers))
```

However, neither of these approaches seem particularly clear to me, since they rely on behavior of the `target.headers` `Set`. I'm slightly surprised that there is not a method to perform the negation of a diff. That would work perfectly here. Instead, we're applying DeMorgan's law.